### DISCUSSION TITLE
Dispatcher initialization at a 5000ms time frame

### DISCUSSION THREAD

**Discussion by VincentVanTongeren** [salesforce/akita#717] [General]
What is the reasoning behind the 5000ms time frame for adding a store?

In dispatcher.ts there is the following line: 
export const $$addStore = new ReplaySubject<string>(50, 5000);

In my use case, I have a custom storage set up with persist state. GetItem is an observable that internally needs to wait until there is a connection with another system in order to retrieve the persisted initial data. This requires a manual action and since the timer starts ticking automatically, more often than not, this 5000ms window passes.

In my local environment, I have removed the 5000ms restriction to the default restriction (Infinity) and now it works like a charm. I would like to know if the 5000ms is there for a very important reason and if I can remove the restriction. The most ideal solution would be to be able to override this parameter or supply my own subject.

**Comment by NetanelBasal** [comment]
I'm not sure how this internal function is related to what you're trying to do.

**Reply by VincentVanTongeren** [reply]
Thanks for getting back to me!

The data is encrypted with a key and that key can only be retrieved when the user is logged in.

The way I understand the code:
- The JavaScript is interpreted and $$addStore has a replaySubject that will expire in 5000ms
- I tell Akita to persist 2 particular stores
- At around 150ms, the store I want to persist is created and $$addStore emits (now the 5000ms are ticking)
- The data from that store is retrieved by calling my getItem (from the PersistStateStorage-interface)
- My implementation includes an observable that emits when someone is logged in
- My data is decrypted and the value is emitted
- The store’s initial value is set

When it takes more than 5000ms to emit the value, $addStore does not emit anymore and the value is not set and the store is loading indefinitely. When I remove the 5000ms from the replaySubject, everything works fine

**Reply by NetanelBasal** [reply]
Can you create a simple example on stackblitz to check what can be done to help?

**Reply by VincentVanTongeren** [reply]
I will, but that will be done tomorrow (Monday)

**Reply by VincentVanTongerenTopicus** [reply]
https://stackblitz.com/edit/angular-ivy-hj94ly?file=src/app/app.component.ts

I have added comments where applicable. I have been able to reproduce the wanted and unwanted behaviour. Make sure to fully refresh though. Uncomment line 32 to make it work as expected.

This shows that the initialization of $addStore in dispatcher.ts is causing problems for this use case. I guess it might have been a micro-optimization, but I hope the optimization can be removed for this use case.

Note that the very first run works as expected. This first run will add a value to the store. After that, the behaviour is not expected anymore. The store is loading, while there should have been data available.

**Reply by NetanelBasal** [reply]
I understand now. But what's the reason for doing this work inside the getItem storage?

**Reply by VincentVanTongerenTopicus** [reply]
We store our sensitive data encrypted with a key that can only be retrieved when you are logged in to our api. So part of the work is waiting for the user to be logged in correctly. All our dependencies to make retrieving and using data happen are loaded when our app starts and are not resolved lazily.

**Reply by NetanelBasal** [reply]
I see. I can increase the buffer time, but if I see that it causes unexpected behavior for users for some reason, I'll revert it.

**Reply by VincentVanTongerenTopicus** [reply]
Thanks for helping us out, we really appreciate it, but this will not solve it for us, because we totally depend on our users to log in, which can take 3 seconds, 5 minutes or an hour. I think we will have to add some complexity setting up our stores after we make sure we are connected and getitem can be finished.

**Reply by NetanelBasal** [reply]
Why not doing it outside of the getItem function? I mean the logic of the login part and decrypt.

**Reply by VincentVanTongerenTopicus** [reply]
Thanks for thinking along with us. As far as I understand, that would require a new hook in akita to process retrieved persisted data, since the encrypted data cannot be mapped to the stores contents directly. I can only supply a method that retrieves data somehow and I assume that the getItem interface on PersistStateStorage expects finalized data that can be used at once.

**Reply by NetanelBasal** [reply]
Hmmm what about something like:

```ts
@Injectable()
export class TodosQuery extends QueryEntity<TodosState> {
  constructor(protected store: TodosStore) {
    super(store);
  }

  select() {
    return super.select().pipe(filter(state => state.decrypted));
  }
}
```

**Reply by NetanelBasal** [reply]
Check this https://stackblitz.com/edit/angular-ivy-fyxaha?file=src%2Fapp%2Fstore%2Ftodos.query.ts

**Reply by VincentVanTongerenTopicus** [reply]
That might work, I will explore this in our project!

Thanks for your time and thoughts!

**Reply by MarekCikos** [reply]
Hello, I have a similar problem.

In dispatcher.ts there is the following line:
`export const $$addStore = new ReplaySubject(50, 5000);`

and in persistState.ts there is the following piece of code in method persistState():
```
subscriptions.push(
	$$addStore.subscribe((storeName) => {
                // ...
		// ...some piece of unimportant code
                // ...
                // creating a subscription to changes in the store
		subscribe(storeName, path);
	})
);
```

Questions:
 1. does this mean that if an application has for example 54 stores, a subscription to changes will be created only for the last 50 stores (because of ReplaySubject capacity of 50) and changes for the first 4 stores will not be stored in the storage (localStorage/DB)?
 2. Does this mean that if there is a time at least 5 seconds between the store constructor call and the persistState method call, `$$addStore.subscribe` will have no values to subscribe and therefore no values will be stored in the storage (localStorage/DB)?

In my use case I have the login page where the user needs to log in. During login page opening, all stores are created. Then the user can log in and this action can take from seconds to minutes or hours (opened application in login page). After the user is logged in, the code calls persistState method and provides custom storage (DB Storage) to save all user's store values to database. After the logout, the storage object from persistState method is destroyed (`storage.destroy()`) and login page appears again for another user to log in. Every user has different stored values and these values are sensitive (store must be clean after every logout and before every login).

My problem was that if the login page was opened for 5 seconds or more, the user's store values were not updating after the login because the subscription was not created (see code above).

I solved this problem in a way that i cloned akita library and rewrote `export const $$addStore = new ReplaySubject(50, 5000);` to `export const $$addStore = new ReplaySubject();` This resolved all my problems. Is there a way to configure these values or can be this feature added to any future releases? Or is there any other way to resolve this kind of problem (long time window between the store creation and the persistState method call)?

Thank you for your advice @NetanelBasal 
Marek
